newer workspace

gohsumin opened the pull request:
### **Better organization**: workspace has its own folder now, and the files are separated into subfolders in a logical way.

<img width="449" alt="Screen Shot 2020-10-15 at 7 16 00 PM" src="https://user-images.githubusercontent.com/49374067/96195496-e7f39a00-0f1a-11eb-903b-363e62a8dede.png">

---------------------

### **Flexible dimensions through Contexts**: the old globals files are rewritten as React Contexts so that the global values can readily reflect the width and height passed into each instance of the workspace.

- The context provider reads the values passed into the workspace as props and makes those values available to each component, such as DrawArrow (which needs to be renamed, now that I'm looking at it).
<img width="529" alt="Screen Shot 2020-10-15 at 7 21 34 PM" src="https://user-images.githubusercontent.com/49374067/96195769-a9121400-0f1b-11eb-895f-2ae8d096d760.png">

- Normally I would've had to nest all of the contexts that each React component uses, but I found a way to work around that from a youtube video. ProviderComposer nests all of the contexts for you and returns a single context provider.
<img width="510" alt="Screen Shot 2020-10-15 at 7 27 36 PM" src="https://user-images.githubusercontent.com/49374067/96196081-816f7b80-0f1c-11eb-868a-a650fb91a864.png">

- Here is an example of how a component would access the data in the contexts.
<img width="660" alt="Screen Shot 2020-10-15 at 7 30 47 PM" src="https://user-images.githubusercontent.com/49374067/96196254-f2169800-0f1c-11eb-9d2f-24807e569df7.png">

- Thanks to the flexible dimensions, the workspace can now be embedded anywhere with different sizes, like on the tutorials page.
<img width="1440" alt="Screen Shot 2020-10-15 at 7 38 52 PM" src="https://user-images.githubusercontent.com/49374067/96196725-2474c500-0f1e-11eb-9bf3-b117033b30cc.png">

---------------------

### **Slightly improved aesthetic of the pop-up canvas**: it is no longer broken.

<img width="1439" alt="Screen Shot 2020-10-15 at 7 32 17 PM" src="https://user-images.githubusercontent.com/49374067/96196395-41f55f00-0f1d-11eb-8cd6-ff6f50938bde.png">

---------------------

### **There is no longer a problem with displaying non-React elements.**

- For example, HTML Input elements can now be used without worrying about errors in positioning. Unfortunately the input itself stopped working, but I am working on that.
<img width="802" alt="Screen Shot 2020-10-15 at 7 41 40 PM" src="https://user-images.githubusercontent.com/49374067/96196885-8af9e300-0f1e-11eb-8e1c-6f6c64ec9fcb.png">

---------------------

### **Settings tab**: this tab will prevent cluttering the workspace with buttons. The aesthetics will be improved in the near future. The theme icon changes along with the theme.

<img width="1440" alt="Screen Shot 2020-10-15 at 7 44 02 PM" src="https://user-images.githubusercontent.com/49374067/96197103-13788380-0f1f-11eb-9950-40af13f2fc09.png">

Commit:
merged with tutorial stuff

Commit:
merging from main

Merge branch 'main' of https://github.com/GlimmerLabs/MIST into sumin

Commit:
brought in changes from main

Merge branch 'main' of https://github.com/GlimmerLabs/MIST into sumin

Commit:
Merge branch 'main' of https://github.com/GlimmerLabs/MIST into sumin
pulling changes from main

Commit:
started using contexts for globals. Got to read the header height in app.js.

Commit:
changed menu; fixed more variables

Commit:
nodes use context now

Commit:
some touches to the look

Commit:
started adding the old menu as well

Commit:
started to work with the custom input field; seeing the light at the end of the tunnel

Commit:
custom bar is in the right place now

Commit:
profiles

Commit:
fixed all the dimensions; pop up canvas image doesn't show; the input box expands the screen horizontally; values and functions in the menu are not animated

Commit:
merged with main

Commit:
merge with sumin

Commit:
merged with sumin branch

Commit:
added buttons to settings tab in ws

Commit:
merge

Commit:
merge

Commit:
Merge branch 'main' into temp

lindavin committed:
Move misplaced code in routes to the right space

lindavin committed:
Add alternative way to connect to MongoAtlas and include print messages that will show which was connected to.

lindavin committed:
Add source

gohsumin committed:
Merge pull request #5 from GlimmerLabs/sumin-mongodb

Sumin mongodb

Commit:
merged with davin's edits

lindavin commented on frontend/src/pages/tutorial.js:
```diff
@@ -263,7 +263,10 @@ function Tutorials() {
                           <SubsectionButton id={subsection.id} type="text" />
                           <SubsectionButton id={subsection.id} type="video" />
                           <SubsectionButton id={subsection.id} type="final" />
-                          <SubsectionButton id={subsection.id} type="challenges"/>
+                          <SubsectionButton
+                            id={subsection.id}
+                            type="challenges"
+                          />
```
There is an inconsistency here.

lindavin commented on frontend/src/workspace/buildingtools/FunNode.js:
```diff
@@ -111,52 +120,36 @@ function FunNode(props) {
-  }, [props.renderFunction]); //Re-renders whenever the render function changes
+  // +----------------------------+
+  // | Trashcan                   |
+  // +----------------------------+------------------------------------
```
There is a set of redundant Trashcan comments here. If this was intended can you make these more different?

lindavin commented on frontend/src/workspace/buildingtools/FunNode.js:
```diff
@@ -78,15 +79,23 @@ function FunNode(props) {
   const y = props.y;
```
Since you are only exporting FunNode, which is particularly large, I think you can just do 
`export default function FunNode(props){....}` on line 75.

lindavin commented on frontend/src/workspace/buildingtools/FunNode.js:
```diff
@@ -52,20 +52,21 @@
 import Portal from "./Portal";
-import gui from "./mistgui-globals.js";
+import gui from "../globals/mistgui-globals.js";
 import MISTImage from "./MISTImage";
 import useImage from "use-image";
-import nodeDimensions from "./globals-nodes-dimensions.js";
-import globals from "./globals.js";
+import { nodeContext } from "../globals/globals-nodes-dimensions.js";
+import { globalContext } from "../globals/global-context";
+import globals from "../globals/globals.js";
```
Since you don't seem to use Portal, MISTImage, or globals, you should remove them.

lindavin commented on frontend/src/workspace/buildingtools/MISTImageCreate.js:
```diff
@@ -0,0 +1,78 @@
+  useEffect(() => {
+    if (animator) {
+      animator.setResolution((props.width < 150) ? props.width : 150,
+      (props.height < 150) ? props.height : 150);
+      animator.frame();
+    }
+  }, [animator, props.width, props.height]);
+
+  return (
+    <canvas
+      className="mist-image"
+      width={props.width}
+      height={props.height}
+      style={{
+        position: "absolute",
+        top: props.y,
+        left: props.x,
+        width: props.width,
+        height: props.height,
+      }}
+      onMouseOut={() => {
+        if (!props.automated) {
+          if (animator) {
+            animator.stop();
+          }
+        }
+      }}
+      onMouseOver={() => {
+        if (!props.automated) {
+          if (animator) {
+            animator.stop();
+            animator.start();
+          }
+        }
+      }}
+      onClick={props.onClick}
+      ref={canvas}
+    />
+  );
+}
```
It looks like there are three files containing MISTImage that is in the codebase. There's MISTImage in MISTImageGallery.js, MISTImage.js, and thie MISTImageCreate.js. Files do not seem to import MISTImage from MISTImageCreate.js. Is there a reason to this or is the plan to change every import for MISTImage to import form MISTImageCreate.js.

lindavin commented on frontend/src/workspace/buildingtools/RenderBox.js:
```diff
@@ -0,0 +1,25 @@
+import React, {useContext} from "react";
+import MISTImage from "./MISTImage";
```
So it looks like this should be from MISTImageCreate rather than MISTImage.

lindavin commented on frontend/src/workspace/buildingtools/MISTImageCreate.js:
```diff
@@ -0,0 +1,78 @@
+import React, { Component, useEffect, useState } from "react";
```
Component is unused.

lindavin commented on frontend/src/workspace/buildingtools/FunNode.js:
```diff
@@ -52,20 +52,21 @@
 // +----------------------------+
 
-import React, { useState, useEffect, useRef } from "react";
+import React, { useState, useEffect, useRef, useContext } from "react";
```
useEffect is not used.

lindavin commented on frontend/src/workspace/buildingtools/ValNode.js:
```diff
@@ -54,15 +54,16 @@
 
-import React, { useState, useRef } from "react";
+import React, { useState, useRef, useContext, useEffect } from "react";
 import { Rect, Group, Text, Image } from "react-konva";
 import Konva from "konva";
 import Portal from "./Portal";
-import gui from "./mistgui-globals.js";
+import gui from "../globals/mistgui-globals.js";
 import MISTImage from "./MISTImage";
```
useEffect and MISTImage were not used.

lindavin commented on frontend/src/workspace/buildingtools/line.js:
```diff
@@ -0,0 +1,67 @@
+import React, { useState, useContext } from "react";
+import { Line, Group, Text, Image } from "react-konva";
```
Text is not used.

lindavin commented on frontend/src/workspace/funbar/FunBar.js:
```diff
@@ -0,0 +1,187 @@
+  function BarBase(props) {
+    return (
+      <Group>
+        {/* <Rect // Blue bar at bottom of workspace
+          x={0}
+          y={0}
+          width={width}
+          height={funBarHeight}
+          fill={props.bg}
+        /> */}
```
If this commented block is not needed, you should delete it or leave a reason as to why it is being kept.

lindavin commented on frontend/src/workspace/funbar/FunBar.js:
```diff
@@ -0,0 +1,187 @@
+            <animated.Rect // function button
+              {...props}
+              width={funbarDimensions.imageButtonWidth}
+              height={funbarDimensions.imageButtonHeight}
+              //stroke={"#424874"}
+              cornerRadius={8}
+              shadowBlur={5}
+              shadowOffset={{ x: 2, y: 3 }}
+              shadowOpacity={0.5}
+              opacity={renderFunction.isRenderable ? 1 : 0.8}
+            />
+          )}
+        </Spring>
+        <Text
+          text={"Image"}
+          width={funbarDimensions.imageButtonWidth}
+          height={funbarDimensions.imageButtonHeight}
+          align={"center"}
+          verticalAlign={"middle"}
+          fill={imageButtonHovered ? "gray" : "white"}
+          fontSize={funBarFontSize}
+          onClick={() => {
+            if (props.renderFunction.isRenderable) {
+              //setImageButtonClicked(true);
+              setImageButtonHovered(false);
```
Line 151 and line 170. Are these being left as an example or are they no longer necessary? If the latter, then the commented line should be removed.

lindavin commented on frontend/src/workspace/funbar/PopupCanvas.js:
```diff
@@ -0,0 +1,217 @@
+    const [imageExists, setImageExists] = useState("initial");
+
+    function SaveImage() {
+      //var newName = props.imageName;
+      //newName = removeOuterWhiteSpace(newName);
+      //var response = getImageExists(newName);
```
These don't seem necessary anymore. If that is the case, you should remove them.

lindavin commented on frontend/src/workspace/funbar/PopupCanvas.js:
```diff
@@ -0,0 +1,217 @@
+              border: "2px solid #008CBA",
+              textAlign: 'center',
+            }}
+            //onClick={}
```
Perhaps a forgotten commented line?

lindavin commented on frontend/src/workspace/globals/globals-funbar-dimensions.js:
```diff
@@ -0,0 +1,41 @@
+  const imageButtonWidth = functionButtonWidth;
+  const imageButtonHeight = props.funBarHeight * 0.5;
+
+  // width = margin + rfTextAreaWidth + margin + functionButtonWidth + margin + imageButtonWidth + margin
```
Unexplained commented line.

lindavin commented on frontend/src/workspace/index.js:
```diff
@@ -993,193 +1000,372 @@ class WorkspaceComponent extends Component {
-                      node.renderFunction.isRenderable
-                        ? node.renderFunction.renderFunction
-                        : false
+          {/* <div onClick={()=>this.deleteWorkspaces('a').then(alert).catch(alert)}>Delete ws</div> */}
```
Here is a dangling comment.

lindavin reviewed (commented):
This was a really nice job. 

1. From a preliminary review, I mostly marked places where there are left over commented out code that were not explained. There was one case of formatting inconsistency. I marked all these places a comment. 

2. I also noticed that there are three files from which MISTImage is exported. These are MISTImage.js, MISTImageGalley.js, and MISTImageCreate.js. Is there a reason that all three are needed?

3. Many of files were missing documentation explaining what their explicit purposes were such as Funbar.js, Popupcanvas.js, Menu1.js, Menu2.js, etc.... Will there be or is there a separate document that discusses how these components fit together?

4. The image on the Popup canvas does not display properly. This is what I get regardless of the function:
![image](https://user-images.githubusercontent.com/55451477/96207207-a83bab00-0f38-11eb-947b-ef03c22790a4.png)

lindavin reviewed (changes requested):
The buttons, such as save image. save/delete workspace, etc... should be functioning like in the current main branch.

lindavin committed:
Preliminary work to add rqeuests

lindavin committed:
New file containing HTTP requests from the workspace

lindavin committed:
Rewriting image related queries in a clearer way.

We rewrite the image related queries following the JSend API response format.
https://github.com/omniti-labs/jsend. We also provide a generic function that
dispatches responses from the server following the JSend API response format.
We also wrap authentication checking into one function so as to make it less
tedious.

lindavin committed:
Provide documented fetch requests in a separate file for modularity

lindavin committed:
Provide testing buttons to demonstrate the fetch requests in http.workspace.js

These buttons are to test that the requests are working properly and to be an
example of what could be done for buttons in the GUI Workspace. These testing
buttons should be removed when the corresponding actions in the GUI Workspace
have been implemented.

Commit:
edits according to Davin's comments

Commit:
Merge branch 'sumin-workspace-backend' of https://github.com/GlimmerLabs/MIST into sumin

Commit:
connected backend to save image and save ws buttons

Commit:
commented out what's malfunctioning in SavedItems in Menu2

lindavin committed:
Merge branch 'main' into sumin

lindavin committed:
Remove unnecessary height checking

lindavin committed:
Patch UserContext not being detected

The UserContext is not being detected by children within the Workspace Component. The current fix is to provide it again within the ContextProvider.

lindavin committed:
STUB save workspace and delete workspace for future refactoring

lindavin committed:
Update gitignore to ignore the build folder

lindavin committed:
Update comment

lindavin committed:
Fix save image button and add stub for saveworkspace

lindavin committed:
Add tap event to switch between tabs

lindavin committed:
Implement touchdrag new line

lindavin committed:
Allow touch to connect value to outlets

lindavin committed:
Allow funnodes to create lines

lindavin committed:
Add touch controls for line creation and for renderbox toggle

lindavin committed:
Merge branch 'sumin' of https://github.com/GlimmerLabs/MIST into sumin

lindavin committed:
Format document, remove unused imports, and fix konva bridging contexts

lindavin committed:
Format document

lindavin committed:
Add popup modal to accept user input to save workspace

lindavin committed:
Fix displaying saved workspaces

lindavin committed:
Fix placeholder

lindavin committed:
Remove unused workspace handlers

lindavin committed:
Add formmodal for deleting workspaces

lindavin committed:
Connect delete popup modal to the menu

lindavin committed:
Connect deleting to the backend

lindavin committed:
Allow deletemodal to close via the close button

lindavin committed:
Add confirmation popup to trigger warning before overwriting

lindavin committed:
Add tap listeners to buttons

lindavin committed:
Remove unnecessary deepclone

lindavin committed:
Add delete touch listeners on edges and nodes

Commit:
createLayout fixed somewhat

lindavin committed:
Change touch listener for deleting a node

lindavin committed:
Remove redundant files from menu

lindavin committed:
Remove unused imports from the main workspace component

lindavin committed:
Rename Menu2 to Menu

lindavin committed:
Modify openWorkspace function

lindavin committed:
Rename buildingtools to buildingTools to follow camelCase

lindavin committed:
Remove unused files from the root level

lindavin committed:
Remove unused imports from App.js

lindavin committed:
Move popcanvas into its own folder

lindavin committed:
Rename PopupCanvas to SaveWorkspaceModal

lindavin committed:
Rename internal use of WorkspacePopupCanvas to SaveWorkspaceModal

lindavin committed:
Rename modal closing handler

lindavin committed:
Rename internal namimng of the show flag of the SaveWorkspace modal

lindavin committed:
Add proptypes for runtime checking to SaveWorkspaceModal

lindavin committed:
Move ConfirmationModal into the modals folder

lindavin committed:
Rename ConfirmationPopup to ConfirmationModal

lindavin committed:
Add propTypes to ConfirmationModal for runtime type checking

lindavin committed:
Move delete workspace modal into the modals folder

lindavin committed:
Complete adding propTypes to and renaming props to modals that connect to the API

lindavin committed:
Rename DeleteWorkspacePopup to DeleteWorkspaceModal and turn openConfimationModal into a class method

lindavin committed:
Move Popupcanvas into modals directory

lindavin committed:
Rename PopupCanvas to ImageModal

lindavin committed:
Remove unused variables and imports, format document, and make ImageModal buttons render dynamically

lindavin committed:
Begin slowly add propTypes to ImageModal

lindavin committed:
Add formatting to Modal references

lindavin committed:
Add propTypes to ImageModal

lindavin committed:
Add key to buttons in ImageModal to resolve React warning

lindavin committed:
Add keys to list items in Menu to resolve React warnings

lindavin committed:
Add keys to menu objects to remove warning

lindavin committed:
Remove unused variables from Custom.js

lindavin committed:
Remove used variables and change variable name to fit camelCase

lindavin committed:
Remove unused import

lindavin committed:
Remove used variables

lindavin committed:
Add proptypes to workspace

lindavin committed:
Change updateHash to not set state directly

lindavin committed:
Resolve warnings for repeated props

lindavin committed:
Remove repeated props and unused variables

lindavin committed:
Remove redundant http.workspace.js file

lindavin committed:
Clean up http.workspace.js which is how we connect the workspace to the API

lindavin committed:
Remove unused files
